[리액트 7기 표후동] 4주차 과제 1 - To-do 리스트 Redux를 사용하여 리팩터링 하기

ndopy opened the pull request:
강의를 보면서 따라해서 구현해보았는데
모든 내용이 완벽하게 이해는 되진 않았지만
'Redux를 이용해서 이렇게 리팩토링이 가능하구나' 정도는 감이 오는 것 같습니다.

반복해서 연습해보겠습니다.

* Coverage의 Uncovered Line 번호
-> ~.test.js 파일의 Line 번호가 아니라 테스트 대상인 ~.jsx 파일의 Line 번호이다.

Commit:
build: React 사용시 import문 생략이 가능하도록 설정한다.

- babel.config.js 파일의 presets 을 다음과 같이 수정한다.
  ㄴ @babel/preset-react 에 runtime: 'automatic' 을 추가한다.

Commit:
refactor: 초기 상태 값을 분리한다.

Commit:
refactor: handleChangeTitle 함수의 상태 변경 코드를 분리한다.

Commit:
refactor: handleClickAddTask 의 상태 변경 코드를 분리한다.

Commit:
refactor: handleClickDeleteTask 함수 내의 상태 변경 코드를 분리한다.

Commit:
refactor: 사용하지 않는 newId 를 제거한다.

Commit:
build: redux 와 react-redux 를 설치했다.

Commit:
feat: store, reducer, Provider 를 추가한다.

Commit:
feat: useSelector 를 추가하고 적용한다.

Commit:
feat: reducer 함수를 작성한다.

Commit:
refactor: useDispatch 를 사용하여 updateTaskTitle 액션을 리팩토링한다.

Commit:
refactor: addTask 액션을 리팩토링한다.

- store.js : reducer 함수에 addTask 액션을 추가한다.
- App.jsx : addTask 함수 내용을 수정한다.

Commit:
refactor: handleClickAddTask 에서 사용하지 않는 state 를 제거한다.

Commit:
refactor: handleClickDeleteTask 함수에서 더 이상 사용하지 않는 state 를 제거한다.

Commit:
refactor: deleteTask 액션을 리팩토링한다.

- store.js : reducer 함수에 deleteTask 액션을 추가한다.
- App.jsx : dispatch 에 전달할 deleteTask 함수를 수정한다.

Commit:
refactor: 상태 초기값을 reducer 함수에 전달한다.

Commit:
refactor: 사용하지 않는 import useState 문을 제거한다.

Commit:
refactor: App.jsx의 action creator들을 actions.js 파일로 분리한다.

Commit:
refactor: useSelector 안의 함수를 selector 함수로 분리한다.

Commit:
feat: jest 에서 사용할 mock 함수를 생성한다.

- __mocks__ 폴더를 생성하고 그 안에 react-redux.js 파일을 생성한다.
- react.redux.js 파일안에 useDispatch와 useSelector 함수를 jest.fn() 으로 생성하고 export 한다.
- 테스트하려는 코드에서 jest.mock('파일명') 으로 mock 함수들을 검색해서 사용할 수 있다.

Commit:
refactor: App.test.jsx 에서 react-redux.js 에 작성한 mock 함수를 사용한다.

- jest.mock('파일명') : 전달한 파일명에서 mock 함수를 검색한다.
  ex) jest.mock('react-redux') : 'react-redux' 라는 파일에서 mock 함수를 검색해서 사용할 수 있다.

- .mockImplementation : mock 함수를 구현해서 사용할 수 있다.
  ㄴ useSelector.mockImplementation : useSelector 라는 mock 함수를 구현해서 사용한다.

Commit:
refactor: reducer 함수를 store.js 로부터 분리한다.

Commit:
chore: reducer.js 안의 불필요한 import문을 제거한다.

Commit:
test: reducer 에 대한 첫 번째 테스트를 작성했다.

- updateTaskTitle 에 대한 테스트 코드를 작성했다.

Commit:
refactor: updateTaskTitle 테스트 코드를 리팩토링했다.

- actions.js 로부터 updateTaskTitle 을 import 한다.
- 기존에 작성한 actions 를 import 한 updateTaskTitle 로 대체한다.
- 더 이상 사용하지 않는 const actions 를 제거한다.

Commit:
fix: addTask 함수 내의 오탈자를 수정한다.

- addTask 함수 내에 type: 'updateTaskTitle' 로 잘못 입력된 내용을 'addTask' 로 수정했다.

Commit:
test: addTask 액션에 대한 테스트를 작성했다.

Commit:
test: addTask 함수에 대한 테스트 코드에 내용을 추가했다.

- 'New Task' 라고 추가한 title이 tasks 배열 안에 있는지 확인하는 코드를 추가했다.

Commit:
test: addTask 에 대한 테스트 코드를 추가했다.

- addTask 이후에 taskTitle 을 지우는(공백으로 만드는) 테스트를 추가했다.

Commit:
refactor: test 코드를 describe 로 구분해준다.

Commit:
test: 할 일 입력을 하지 않은 상태로 addTask 를 호출할 때의 테스트 코드를 추가한다.

- 'without task title' 이라는 context 로 구분한다.
- reducer 함수의 addTask 에서 taskTitle 이 없을 때 처리하는 코드를 추가한다.

Commit:
test: addTask 테스트에 id 에 대한 테스트를 추가한다.

- newId 라는 프로퍼티를 추가한다.
- id 값이 undefined 가 아닌지를 테스트한다.

Commit:
refactor: addTask 테스트 코드 리팩토링

- 공통으로 사용하는 코드를 reduceAddTask 라는 함수로 분리했다.

Commit:
test: deleteTask 에 대한 테스트 코드를 추가했다.

- 기존제 존재하는 ID 의 task를 삭제할 때 : tasks 의 길이가 0이어야 한다.
- 기존에 없는 ID의  task를 삭제할 때 : 동작하지 않는다.

Commit:
refactor: App.jsx 에서 ListContainer 로 분리한다.

- <ListContainer> 컴포넌트를 새로 생성한다.
- ListContainer 컴포넌트의 관심사는 할일 목록에서 할일이 제거되는 것이다.
  ㄴ 그 외는 input, button 의 관심사이다.

- Page.jsx: List 대신 ListContainer 로 대체한다.
  ㄴ List.jsx 에서는 화면에 보여주는 것만을 담당한다.
  ㄴ 동작은 ListContainer.jsx 에서 담당한다.

Commit:
test: ListContainer 컴포넌트에 대한 테스트 코드를 작성했다.

- 화면에 호출하여 '아무것도 하지 않기 #1'이 존재하는지 확인한다.

Commit:
build: .eslintrc.js 파일 내용에 옵션을 추가했다.

- ListContainer.test.js 파일 내에서 <ListContainer /> 를 작성할 때
  JSX not allowed in files with extension ' .js' 라는 에러가 발생했다.
  ㄴ 'react/jsx-filename-extension': [0], 옵션을 추가하였고, 해당 에러 메시지는 사라졌다.

Commit:
refactor: Page.test.js 리팩토링

- useSelector 와 jest.mock 코드를 사용한다.
- 사용하지 않는 부분을 제거한다.

Commit:
test: InputContainer 에 대한 테스트 코드를 작성했다.

Commit:
test: InputContainer 에 테스트 코드를 추가로 작성했다.

- getByDisplayValue 를 이용한 테스트 코드를 추가했다.

Commit:
refactor: Page 에 대한 코드들을 리팩토링한다.

- Page.jsx : 사용하지 않는 <Input/> 대신 <InputContainer/> 를 작성한다.
- Page.test.jsx : 사용하지 않는 코드들을 제거한다.

Commit:
refactor: Page.jsx의 내용을 App.jsx 로 옮긴다.

- Page.jsx 는 단순히 화면에 컴포넌트를 보여주는 기능을 하기 때문에
  App.jsx 에서 이를 수행해도 문제될 것이 없다.

- Page.test.jsx 에 작성된 테스트 내용은 App.test.jsx 에 옮긴다.

- 더 이상 사용하지 않는 Page.jsx, Page.test.jsx 는 제거한다.

moonkii commented on src/InputContainer.jsx:
```diff
@@ -0,0 +1,32 @@
+    taskTitle: state.taskTitle,
+  }));
+
+  function handleChangeTitle(event) {
+    dispatch(updateTaskTitle(event.target.value));
+  }
```
`event` 객체를 받은 형태 보다는 값을 전달받는 형태로하면 의존성을 끊어줄 수 있을거에요 ;)

ndopy replied:
이전에 알려주신 방법을 적용안하고 있었네요.. 적용해보겠습니다!

Commit:
test: InputContainer 테스트 코드를 추가했다.

- changes task title 에 대한 테스트 코드를 작성했다.
  ㄴ input 에 '새로운 할 일' 이라고 입력했을 때 updateTaskTitle 이 호출되는지?

moonkii commented on src/InputContainer.test.js:
```diff
@@ -0,0 +1,31 @@
+
+  it('shows task title ', () => {
+    useSelector.mockImplementation((selector) => selector({
+      taskTitle: 'New Title',
```
각각 테스트는 독립적으로 시행되는 것이기 때문에 redux 부분을 `beforeEach` 로 빼주시는게 좋아요!

moonkii commented on src/InputContainer.test.js:
```diff
@@ -0,0 +1,31 @@
+jest.mock('react-redux');
+
+describe('InputContainer', () => {
+  const dispatch = jest.fn();
```
테스트 시나리오가 하나긴 하지만 저는 가능하면 mock function 이 있는경우 초기화를 해주는 편이에요.

ndopy replied:
각 테스트가 끝날 때마다 afterEach 로 `jest.clearAllMocks()` 를 호출하는 방식으로 작성해보려합니다.

moonkii commented on src/ListContainer.test.js:
```diff
@@ -0,0 +1,26 @@
+      tasks,
+    }));
+
+    const { getByText } = render((
```
`getBy~` 와 `queryBy~` 를 구분해서 사용하시면 좋아요.

ndopy replied:
`getBy~` 와 `queryBy~` 의 큰 차이점은 
'조건에 맞는 엘리먼트가 검색되지 않는다면 에러를 발생시키는가? 아니면 null 을 반환하는가?'로 파악했습니다.

이전에 클린 코드에 관련된 내용 중에서 'null 을 반환하지 말라'라는 내용을 본 적이 있습니다.
https://eglowc.tistory.com/42 

물론 Java 와 관련하여 작성된 내용이라 자바스크립트와는 다를수도 있겠다?는 생각이 들지만
동작하지 않을 때 null 대신 오히려 에러를 반환하는 것이
코드를 작성하는 개발자에게 더 확실히 잘못됐다는 것을 알리는? 그런 느낌이라서
`getBy~` 를 쓰는 것이 더 맞다고 생각이 듭니다.

`getBy~` 와 `queryBy~` 의 사용처를 어떻게 구분해야 할지는 아직 이해를 못하고 있는 것 같습니다. 😓

moonkii replied:
차이는 잘 찾으셨어요!

직관적으로 생각해보시면 바로 답을 찾으실거에요.
A 라는 텍스트가 있느냐, 없느냐 를 테스트하면 결과는 있다, 없다로 나올거에요.
그런데 `getBy~` 를 사용하면 에러가 발생해버리죠.
A 라는 텍스트가 없는게 에러 상황인가? 생각해보면 아닐거에요.

그러면 언제 `getBy~` 를 사용할 수 있을까 고민해보면
예를들어 버튼 클릭 이벤트를 테스트할 때는 이미 버튼이 있다는 전제하에 테스트가 되어야할거에요.
그렇다면 버튼이 없는 경우는 에러가 발생해야되겠죠? 없는 버튼의 클릭 이벤트를 발생시킬 수가 없으니까요.

운전자에게 주의를 주기 위해 경보를 울리지만 경찰차, 소방차, 구급차 등이 다 다른 경보를 사용하죠.
마찬가지로 프로그래머가 버그를 발견하기 위해 테스트 코드를 짜고 잘못된 부분에 대해 알림을 받는 것은 맞지만
의미가 맞아야 된다고 생각해요. 사소한 부분이지만 제 3자의 입장에서 코드를 읽을 때는 큰 차이로 다가올 수 있죠.

클린코드를 읽으면서 '왜' `null` 을 사용하지 말라고 했는지 이유를 고민해보셨나요?
여기서는 왜 `null`을 반환하면 좋지 않다고 생각하셨나요?

ndopy replied:
> 그러면 언제 getBy~ 를 사용할 수 있을까 고민해보면
예를들어 버튼 클릭 이벤트를 테스트할 때는 이미 버튼이 있다는 전제하에 테스트가 되어야할거에요.
그렇다면 버튼이 없는 경우는 에러가 발생해야되겠죠? 없는 버튼의 클릭 이벤트를 발생시킬 수가 없으니까요.

여기에서 이해가 확 와닿았습니다. 👍 

---

> 클린코드를 읽으면서 '왜' null 을 사용하지 말라고 했는지 이유를 고민해보셨나요?

다시 읽어보니 클린 코드에서 null 을 사용하지 말라고 한 이유는
null 을 반환한 뒤에 전달받은 곳에서 다음과 같이 확인해야 하는 코드를 작성해야 하기 때문이라네요.
```javascript 
if (object != null)
```
이렇게 보면 지금 작성하는 테스트 코드와는 거리가 먼 내용같네요.

---

> 여기서는 왜 null을 반환하면 좋지 않다고 생각하셨나요?

사실 '클린 코드에서 null 반환을 하면 안된다.'라고 해서 맹목적인 믿음(?)으로 따라했습니다. 😅 
이제부터라도 "왜 그렇게 코드를 작성했는가?"라는 이유를 계속 고민해 보겠습니다.

Commit:
refactor: InputContainer 공통 사용 코드를 함수로 분리한다.

moonkii commented on src/reducer.js:
```diff
@@ -0,0 +1,44 @@
+// - type (string)
+// - payload => object = { taskTitle, ... }
+
+export default function reducer(state = initialState, action) {
```
매개변수를 destructuring 해주시면 사용하실 때 더 편하실거에요.

ndopy replied:
매개변수 `state` 를 구조 분해 할당하면 다음과 같이 작성할 수 있을 것 같습니다.
![image](https://user-images.githubusercontent.com/55405713/162794050-086f876a-7d96-4e4e-8ca1-8b3bc970daed.png)

그런데 이렇게 되면 `...state` 대신 `{ newId, taskTitle, tasks }`  와 같이 작성해야하는데
오히려 더 작성해야 할 내용이 많아지는 것이 아닐까 하는 생각이 들었습니다.
아니면 제가 요점을 잘못 파악하고 있는 것일 것 같기도 합니다... 😢

moonkii replied:
저는 `reducer({ state, action })` 형태를 의미한 건데 이 부분은 코드를 작성해보시고 본인에게 더 편하거나 명확하다고 생각드는 쪽을 선택하시면 될 것 같아요.
이렇게 해야한다라는 정답을 말하려고 하는게 아니라 '이렇게 해보면 어떨까요?' 피드백을 드리는 것이기 때문에 항상 스스로 고민해보시고 적용하시는게 중요해요 ;)

moonkii commented on src/reducer.test.js:
```diff
@@ -0,0 +1,87 @@
+        expect(state.taskTitle).toBe('');
+      });
+
+      context('without task title', () => {
+        it("doesn't work", () => {
+          const state = reduceAddTask();
+
+          expect(state.tasks).toHaveLength(0);
+        });
```
이 부분은 depth 를 잘 못잡으신 것 같아요. 대조되는 context 끼리 depth 가 맞아야 할거에요!

ndopy replied:
말씀하신 내용이 맞습니다. 코드를 잘못 위치시켰네요. 수정하겠습니다.

moonkii commented on src/reducer.test.js:
```diff
@@ -0,0 +1,87 @@
+  });
+
+  describe('deleteTask', () => {
+    context('with existed task ID', () => {
+      it('remove the task from tasks', () => {
+        const previousState = {
+          tasks: [
+            { id: 1, title: 'Task' },
+          ],
+        };
+
+        const state = reducer(previousState, deleteTask(1));
+
+        expect(state.tasks).toHaveLength(0);
+      });
+    });
+
+    context('without existed task ID', () => {
```
이럴 때는 when 을 사용해주시면 context 표현이 더 자연스러울 거에요.

ndopy replied:
when 을 이용한 표현으로 적용해보겠습니다.

Commit:
test: ListContainer 컴포넌트 테스트 코드를 추가했다.

- tasks 가 빈 배열일 때 '할 일이 없어요!' 라는 메시지가 보이는지 테스트한다.
- 공통으로 사용하는 코드들을 분리한다.

moonkii reviewed (commented):
![image](https://user-images.githubusercontent.com/46776329/162771962-7312b4fb-ec3f-465c-8789-63b258c0e209.png)

전체적으로 깔끔하게 코드를 잘 작성하신 것 같아요. 💯 
이정도 느낌으로 패턴을 외운다 생각하고 하셔도 좋아요.
PR 보내실 때 린트, test 를 로컬에서 항상 확인해보시면 좋아요.
추가적으로 PR 보내시고 나서도 셀프리뷰하는 습관을 들이시면 크게 피드백 받을 게 없을 정도로 코드 작성이 가능하실거에요 ;)

Commit:
style: 코드 포맷팅 변경

- 불필요하다고 생각되는 공백들을 정리하였다.

Commit:
test: ListContainer 컴포넌트의 테스트 코드를 추가했다.

- '완료' 버튼의 이벤트 동작에 대한 테스트를 추가했다.
- 할 일이 없을 경우, 완료 버튼이 없는 것을 확인하는 테스트를 추가했다.

Commit:
chore: 불필요한 import 문을 제거했다.

Commit:
test: reducer 테스트에 action type이 정의되지 않았을 경우를 테스트하는 코드를 추가한다.

- reducer 함수에는 별도의 else 문이 없다.
- action.type에 updateTaskTitle, addTask, deleteTask 도 아닌
  다른 문자열이 올 경우 else에 해당하는 return state; 가 실행된다.

- 따라서 reducer 를 호출할 때 action.type을 빈 문자열('')로 전달하면
  기존의 상태가 그대로 return 될 것이기 때문에
  return 된 state 와 기존 state 가 toEqual 인지 테스트하면 될 것 같았다.

[이슈]
- test coverage 에서 % Branch 가 88.89% 로 100%가 되지 않았다.
  updateTaskTitle, addTask, deleteTask 가 아닌 문자열로 테스트했는데도
  else 테스트가 되지 않는 것인가?

Commit:
refactor: InputContainer 테스트 코드를 리팩토링한다.

- beforeEach 를 이용해서 각 테스트가 시작할 때마다 실행할  Redux 코드를 분리한다.
- afterEach 를 이용해서 각 테스트가 끝날 때마다  mock 함수들을 all clear 한다.

Commit:
refactor: ListContainer 테스트 코드를 리팩토링한다.

- beforeEach 를 이용해서 각 테스트가 시작할 때마다 useDispatch 를 구현한다.
- afterEach 를 이용해서 각 테스트가 끝날 때마다 mock 함수들을 all clear 한다.

Commit:
chore: 'without task title' Context 의 depth 를 수정했다.

- 'with task title' Context 안에 'without task title' 이 작성되어 있었다.
  ㄴ 대조되는 Context 끼리 depth 가 맞아야하기 때문에 코드 위치를 수정했다.

Commit:
chore: reducer 테스트에 대한 context 표현을 수정했다.

- task ID 가 있을 때와 없을 때를 표현해야 해서 when 을 이용한 표현으로 수정했다.
  ㄴ task ID가 있을 때 : when there is a task ID
  ㄴ task ID가 없을 때 : when there is no task ID

Commit:
refactor: event 객체 대신 값을 전달받는 형태로 리팩토링한다.

- 상위 컴포넌트인 InputContainer 에서 전달받은 onChange를 처리할 때
 ㄴ 하위 컴포넌트인 Input 에서 별도의 handleChange 함수를 작성한다.
 ㄴ handleChange 함수의 매개변수로 event 객체를 전달한다.
 ㄴ handleChange 함수 내용에서 event.target.value 를 추출한다. (ex. value: inputTitle)
 ㄴ 상위 컴포넌트에서 전달받은 onChange 를 실행할 때
    event 의 target.value 값인 inputTitle 을 인자로 전달하여 실행한다.

- 이렇게 작성하면 상위 컴포넌트는 단순히 값만 받아서 업데이트하는 구조를 만들 수 있다.

moonkii commented:
> 명시되지 않은 else 에 대한 테스트에 대해 구글링을 해보고 있는데 이해를 못하고 있습니다.
도움이 될만한 키워드가 있을까요?

테스트 결과를 참고하시는게 제일 좋아요.
11번째 줄이 문제라는 것에서 아마 매개변수쪽에 테스트 케이스를 놓치고 있구나라고 유추해볼 수 있죠.
여기서는 initial state 가 적용되는 경우의 테스트 케이스가 빠져서 그럴거에요.

Commit:
test: reducer 함수의 state 초깃값에 대한 테스트를 추가했다.

- reducer 함수를 호출할 때 state 인자를 undefined 로 전달하면 state 의 초깃값이 반환된다.

moonkii commented on src/reducer.test.js:
```diff
@@ -104,4 +104,16 @@ describe('reducer', () => {
     });
   });
+
+  describe('initial value', () => {
```
describe 는 테스트 서술대상이에요.
대상을 it에 넣어서 테스트 시나리오를 읽어보세요.

initial value returns initial state 라고하면 어떻게 해석이되나요?

moonkii commented on src/reducer.test.js:
```diff
@@ -104,4 +104,16 @@ describe('reducer', () => {
+
+  describe('initial value', () => {
+    context('when reducer function called with undefined state ', () => {
+      it('return initial state', () => {
```
it 은 3인칭 단수이기 때문에 문법상 returns 가 될거에요!
